pkg/controller/association/dynamic_watches.go

```diff
@@ -43,17 +43,39 @@ func serviceWatchName(associated types.NamespacedName) string {
 // with each reconciliation.
 func (r *Reconciler) reconcileWatches(associated types.NamespacedName, associations []commonv1.Association) error {
 	// watch the referenced ES cluster for future reconciliations
+	// TODO: what if the referenced resource is not an Elasticsearch resource? https://github.com/elastic/cloud-on-k8s/issues/4591
 	if err := ReconcileWatch(associated, associations, r.watches.ElasticsearchClusters, esWatchName(associated), func(association commonv1.Association) types.NamespacedName {
 		return association.AssociationRef().NamespacedName()
 	}); err != nil {
 		return err
 	}
 
-	// watch the user secret in the ES namespace
-	if err := ReconcileWatch(associated, associations, r.watches.Secrets, esUserWatchName(associated), func(association commonv1.Association) types.NamespacedName {
-		return UserKey(association, association.AssociationRef().Namespace, r.UserSecretSuffix)
-	}); err != nil {
-		return err
+	if r.ElasticsearchUserCreation != nil {
+		// watch the user secret in the ES namespace
+
+		// TODO: this is not great. This reconcileWatches function is called by each association controller.
+		// For example for Kibana, called both by the kb-es and kb-ent controllers.
+		// Both controllers override each other's watches. It's OK when they write the same thing, but in this case
+		// the kb-ent controller does not setup any es user while the kb-es controller does.
+		// Because the context is different in both controller we end up with those weird if conditions.
+		// Things would be more consistent if each association controller sets up its own watches rather than
+		// having multiple association controllers setting up all associations watches for a given resource.
+		// This way we'd just check r.ElasticsearchUserCreation != nil, rather than checking again later
+		// for each association whether it requires auth.
```

Comment:
This is the part I don't like much.
@david-kow I remember we discussed doing things that way when we introduced the agent multi-association support, for simplicity. Now it's causing some troubles since multiple association controllers (kb-es, kb-ent) reconcile the same watches with a different reconciler parameters.


I'm trying to think about a better solution.

Reply:
I ended up refactoring the dynamic watches to make things simpler and more consistent.